contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,587 @@
+    }
+
+    /// @notice Updates a pool and user boost factors after
+    /// a veJoe token operation. This function needs to be called
+    /// by the veJoe contract after every event.
+    /// @param _user The users address we are updating
+    /// @param _newUserBalance The new balance of the users veJoe
+    function updateBoost(address _user, uint256 _newUserBalance) external {
+        require(msg.sender == address(VEJOE), "BoostedMasterChefJoe: Caller not veJOE");
+
+        uint256 length = poolInfo.length;
+
+        for (uint256 pid = 0; pid < length; ++pid) {
+            UserInfo storage user = userInfo[pid][_user];
+
+            // Skip if user doesn't have any deposit in the pool
+            if (user.amount == 0) {
+                continue;
+            }
+
+            PoolInfo storage pool = poolInfo[pid];
+            updatePool(pid);
+            // calculate pending
+            uint256 oldBoostedLiquidity = getBoostedLiquidity(pid, _user);
+            uint256 pending = oldBoostedLiquidity.mul(pool.accJoePerShare).div(ACC_TOKEN_PRECISION).sub(
+                user.rewardDebt
+            );
+
+            // Increase claimableJoe
+            claimableJoe[pid][_user] += pending;
+
+            // Update users veJoeBalance
+            uint256 _oldBalance = user.veJoeBalance;
+            user.veJoeBalance = _newUserBalance;
+
+            // Update the pool total veJoe
+            pool.totalVeJoe = pool.totalVeJoe.add(_newUserBalance).sub(_oldBalance);
+
+            uint256 newBoostedLiquidity = getBoostedLiquidity(pid, _user);
+
+            // Update the pools total boosted
+            pool.totalBoostedAmount = pool.totalBoostedAmount.add(newBoostedLiquidity).sub(oldBoostedLiquidity);
+
+            // Update reward debt, take into account new boost
+            uint256 debt = newBoostedLiquidity.mul(pool.accJoePerShare).div(ACC_TOKEN_PRECISION);
+        }
+    }
+
+    /// @notice Returns a users "boosted" liquidity.
+    /// @param _pid The index of the pool. See `poolInfo`
+    /// @param _user The address of the depositor
+    /// @return boostedLiquidity The users liquidity multiplied by their boost.
+    function getBoostedLiquidity(uint256 _pid, address _user) public view returns (uint256 boostedLiquidity) {
+        UserInfo memory user = userInfo[_pid][_user];
+        boostedLiquidity = user.amount.mul(getUserBoost(_pid, _user)).div(1000);
+    }
+
+    /// @notice Returns a users boosted emission share
+    /// @param _pid The index of the pool. See `poolInfo`
+    /// @param _user The address of the depositor
+    /// @dev `userBoost` is scaled by 1e3
+    /// @return userBoost A multiplier for a users liquidity
+    function getUserBoost(uint256 _pid, address _user) public view returns (uint256 userBoost) {
+        PoolInfo memory pool = poolInfo[_pid];
+        UserInfo memory user = userInfo[_pid][_user];
+        userBoost = 1000;
+        if (user.veJoeBalance > 0 && pool.totalVeJoe > 0) {
```

Comment:
Can do `if (user.veJoeBalance != 0 && pool.totalVeJoe != 0)` instead to save gas